File: 0001-0500/0020/0020.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>();
        foreach (char c in s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.Push(c);
            } else if (stack.Count > 0 && ((c == ')' && stack.Peek() == '(') || (c == '}' && stack.Peek() == '{') || (c == ']' && stack.Peek() == '['))) {
                stack.Pop();
            } else {
                return false;
            }
        }
        return stack.Count == 0;
    }
}
// @lc code=end

File: 0001-0500/0020/0020.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        char c;
        stack<char> st;
        for (int i = 0; i < s.size(); i++)
        {

            switch (s[i])
            {
            case '(':
                st.push(s[i]);
                break;
            case '{':
                st.push(s[i]);
                break;
            case '[':
                st.push(s[i]);
                break;
            case ')':
                if (st.empty())
                {
                    return false;
                }
                if (st.top() != '(')
                {
                    return false;
                }
                st.pop();
                break;
            case '}':
                if (st.empty())
                {
                    return false;
                }
                if (st.top() != '{')
                {
                    return false;
                }
                st.pop();
                break;
            case ']':
                if (st.empty())
                {
                    return false;
                }
                if (st.top() != '[')
                {
                    return false;
                }
                st.pop();
                break;
            default:
                return false;
            }
        }
        return st.empty();
    }
};
// @lc code=end
